test/BIP324: functional tests for v2 P2P encryption

stratospher opened the pull request:
This PR introduces support for v2 P2P encryption(BIP 324) in the existing functional test framework and adds functional tests for the same.

### commits overview
1. introduces a new class `EncryptedP2PState` to store the keys, functions for performing the initial v2 handshake and encryption/decryption.
3. this class is used by `P2PConnection` in inbound/outbound connections to perform the initial v2 handshake before the v1 version handshake. Only after the initial v2 handshake is performed do application layer P2P messages(version, verack etc..) get exchanged. (in a v2 connection)
    - `v2_state` is the object of class `EncryptedP2PState` in `P2PConnection` used to store its keys, session-id etc.
    - a node [advertising](https://github.com/stratospher/blogosphere/blob/main/integration_test_bip324.md#advertising-to-support-v2-p2p) support for  v2 P2P is different from a node actually [supporting v2 P2P](https://github.com/stratospher/blogosphere/blob/main/integration_test_bip324.md#supporting-v2-p2p) (differ when false advertisement of services occur)
        - introduce a boolean variable `supports_v2_p2p` in `P2PConnection` to denote if it supports v2 P2P.
        - introduce a boolean variable `advertises_v2_p2p` to denote whether `P2PConnection` which mimics peer behaviour advertises V2 P2P support. Default option is `False`.
    - In the test framework, you can create Inbound and Outbound connections to `TestNode`
        1. During **Inbound Connections**, `P2PConnection` is the initiator [`TestNode` <--------- `P2PConnection`]
            - Case 1:
                - if the `TestNode` advertises/signals v2 P2P support (means `self.nodes[i]` set up with `"-v2transport=1"`), different behaviour will be exhibited based on whether:
                    1. `P2PConnection` supports v2 P2P
                    2. `P2PConnection` does not support v2 P2P
               - In a real world scenario, the initiator node would intrinsically know if they support v2 P2P based on whatever code they choose to run. However, in the test scenario where we mimic peer behaviour, we have no way of knowing if `P2PConnection` should support v2 P2P or not. So `supports_v2_p2p` boolean variable is used as an option to enable support for v2 P2P in `P2PConnection`.
              - Since the `TestNode` advertises v2 P2P support (using "-v2transport=1"), our initiator `P2PConnection` would send:
                1. (if the `P2PConnection` supports v2 P2P) ellswift + garbage bytes to initiate the connection
                2. (if the `P2PConnection` does not support v2 P2P) version message to initiate the connection
           - Case 2:
                - if the `TestNode` doesn't signal v2 P2P support; `P2PConnection` being the initiator would send version message to initiate a connection.
       2. During **Outbound Connections** [TestNode --------> P2PConnection]
           - initiator `TestNode` would send:
                - (if the `P2PConnection` advertises v2 P2P) ellswift + garbage bytes to initiate the connection
                - (if the `P2PConnection` advertises v2 P2P) version message to initiate the connection
          - Suppose `P2PConnection` advertises v2 P2P support when it actually doesn't support v2 P2P (false advertisement scenario)
               - `TestNode` sends ellswift + garbage bytes
               - `P2PConnection` receives but can't process it and disconnects.
               - `TestNode` then tries using v1 P2P and sends version message
               - `P2PConnection` receives/processes this successfully and they communicate on v1 P2P

4. the encrypted P2P messages follow a different format - 3 byte length + 1-13 byte message_type + payload + 16 byte MAC
5. includes support for testing decoy messages and v2 connection downgrade(using false advertisement - when a v2 node makes an outbound connection to a node which doesn't support v2 but is advertised as v2 by some malicious
intermediary)

### run the tests
* functional test - `test/functional/p2p_v2_encrypted.py` `test/functional/p2p_v2_earlykeyresponse.py`

I'm also super grateful to @ dhruv for his really valuable feedback on this branch.
Also written a more elaborate explanation here - https://github.com/stratospher/blogosphere/blob/main/integration_test_bip324.md

stratospher committed:
[rpc/net] Allow v2 p2p support in addconnection

This test-only RPC is required when a TestNode initiates
an outbound v2 p2p connection. Add a new arg `v2transport`
so that the node can attempt v2 connections.

stratospher committed:
[test] Introduce EncryptedP2PState object in P2PConnection

Instantiate this object when the connection supports v2 P2P transport
protocol.

- When a P2PConnection is opened, perform initiate_v2_handshake() if the
connection is an initiator. application layer messages are only sent after
the initial v2 handshake is over (for both initiator and responder).

stratospher committed:
[test] Perform initial v2 handshake

stratospher committed:
[test] Build v2 P2P messages

stratospher committed:
[test] Allow inbound and outbound connections supporting v2 P2P protocol

- Add an optional `supports_v2_p2p` parameter to specify if the inbound
and outbound connections support v2 P2P protocol.
- In the `addconnection_callback` which gets called when creating
outbound connections, call the `addconnection` RPC with v2 P2P protocol
support enabled.

jonatack commented:
Concept ACK

brunoerg commented:
Concept ACK

stratospher commented:
converting this PR into a draft. I'll push the updated version which includes the new spec changes in BIP 324 soon.

stratospher committed:
[test/crypto] Add ECDH

Co-authored-by: Pieter Wuille <pieter.wuille@gmail.com>

stratospher committed:
[test] Construct class to handle v2 P2P protocol functions

The class `EncryptedP2PState` stores the 4 32-byte keys, session id,
garbage terminators, whether it's an initiator/responder, whether the
initial handshake has been completed etc.. It also contains functions
to perform the v2 handshake and to encrypt/decrypt p2p v2 messages.

- In an inbound connection to TestNode, P2PConnection is the initiator
and `initiate_v2_handshake()`, `complete_handshake()`, `authenticate_handshake()`
are called on it. [ TestNode <----------------- P2PConnection ]

- In an outbound connection from TestNode, P2PConnection is the responder
and `respond_v2_handshake()`, `complete_handshake()`, `authenticate_handshake()`
are called on it. [ TestNode -----------------> P2PConnection ]

stratospher committed:
[test] Add functional tests to test v2 P2P behaviour

stratospher committed:
[test] Add functional test to test early key response behaviour in BIP 324

- A node initiates a v2 connection by sending 64 bytes ellswift
- In BIP 324 "The responder waits until one byte is received which does not match the
  V1_PREFIX (16 bytes consisting of the network magic followed by "version\x00\x00\x00\x00\x00".)"
- It's possible that the 64 bytes ellswift sent by an initiator starts with a prefix of V1_PREFIX
- Example form of 64 bytes ellswift could be:
	4 bytes network magic + 60 bytes which aren't prefixed with remaining V1_PREFIX
- We test this behaviour:
	- when responder receives 4 byte network magic -> no response received by initiator
	- when first mismatch happens -> response received by initiator

stratospher committed:
[test] Reconnect using v1 P2P when v2 P2P terminates due to magic byte mismatch

- When a v2 TestNode makes an outbound connection to a P2PInterface node
which doesn't support v2 but is advertised as v2 by some malicious
intermediary, the TestNode sends 64 bytes ellswift. The v1 node doesn't
understand this and disconnects. Then the v2 TestNode reconnects by
sending a v1/version message.

stratospher committed:
[test] Ignore BIP324 decoy messages

Also allow P2PConnection::send_message() to send decoy messages for
writing tests.

stratospher committed:
[test] Read v2 P2P messages

stratospher commented:
I've updated the PR to reflect the new spec changes in the BIP. It's built on top of #24545's 78c3ccc. Only the last 19 commits belong to this PR.

fanquake commented:
Moved to draft given it's based on multiple other PRs.

stratospher committed:
[test] Check whether v2 TestNode performs downgrading

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -598,6 +598,11 @@ def wait_for_disconnect(self, timeout=60):
         test_function = lambda: not self.is_connected
         self.wait_until(test_function, timeout=timeout, check_connected=False)
 
+    def wait_for_reconnect(self, timeout=60):
```
ideally, first `wait_for_connect` waits on v2 connection and the `wait_for_connect` a few lines below waits for v1 connection.

it's possible that the v2 reconnection happens quickly and the first `wait_for_connect` detects a v2 connection directly. Then we'd be stuck waiting for disconnect and connect (in the lines below) to happen (even though it has already happened).

mzumsande replied:
Maybe it would work just not checking for the racy intermediate states, but instead waiting until we have received the "version" message via v1? Something like

 ```
def test_function():
     if not (self.is_connected and self.last_message.get('version') and self.v2_state is None):
         return False
     return True
self.wait_until(test_function, timeout=timeout, check_connected=False  )
```

stratospher replied:
good idea! that should work.

mzumsande reviewed (commented):
Concept ACK

just started reviewing, some initial comments

stratospher commented on src/net.cpp:
```diff
@@ -1913,7 +1913,11 @@ bool CConnman::AddConnection(const std::string& address, ConnectionType conn_typ
 
-    OpenNetworkConnection(CAddress(), false, std::move(grant), address.c_str(), conn_type, /*use_v2transport=*/false);
+    CAddress addr(CService(), NODE_NONE);
+    if (use_v2transport) {
```
thanks! done.

mzumsande reviewed (commented):
halfway through with the review, these are my comments so far.

mzumsande commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +97,143 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+            return 0, b""
+        enc_contents_len = response[:LENGTH_FIELD_LEN]
+        response = response[LENGTH_FIELD_LEN:]
+        contents_len = int.from_bytes(self.peer['recv_L'].crypt(enc_contents_len), 'little')
```
I have little knowledge of cryptography, but I think that something is wrong here.
Doesn't `crypt` change the internal state of the stream cipher, so if we abort below because we haven't received the message in full, but then call `crypt` again on the same first 3 bytes of the now larger response, we will get an incorrect result?

stratospher replied:
nice catch. you're right!

i've stored `contents_len` in a class variable so as to not perform the decryption of 3 bytes length again. initially, `contents_len` is -1, after the 3 bytes length gets processed - `contents_len` is updated, when the whole packet is processed `contents_len` is reset to -1 again.

i'm open to other solutions.

i wish there was a cleaner way for writing this function with reads like [this](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki?plain=1#L500-L513) but we're working with `P2PConnection`(subclass of `asyncio.Protocol`) where no read is possible(no "waiting" for data to be ready) and we just have a [`data_received()`](https://docs.python.org/3/library/asyncio-protocol.html#asyncio.Protocol.data_received) callback when data is received.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +97,143 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+        # Optionally send decoy packets after garbage terminator.
+        aad = self.sent_garbage
+        for decoy_content_len in [random.randint(1, 100) for _ in range(10)]:
+            msg_to_send += self.v2_enc_packet(decoy_content_len * b'\x00', aad=aad)
```
done. true! logs had lines like these: `[net] non-version message before version handshake. Message "" from peer=0`.

stratospher commented on src/net.h:
```diff
@@ -1209,13 +1209,14 @@ class CConnman
      * @param[in]   address     Address of node to try connecting to
      * @param[in]   conn_type   ConnectionType::OUTBOUND, ConnectionType::BLOCK_RELAY,
      *                          ConnectionType::ADDR_FETCH or ConnectionType::FEELER
+     * @param[in]   use_v2transport  Set to true if node supports BIP324 v2 transport protocol
```
done.

stratospher commented on src/rpc/net.cpp:
```diff
@@ -357,6 +357,7 @@ static RPCHelpMan addconnection()
         {
             {"address", RPCArg::Type::STR, RPCArg::Optional::NO, "The IP address and port to attempt connecting to."},
             {"connection_type", RPCArg::Type::STR, RPCArg::Optional::NO, "Type of connection to open (\"outbound-full-relay\", \"block-relay-only\", \"addr-fetch\" or \"feeler\")."},
+            {"v2transport", RPCArg::Type::BOOL, RPCArg::Default{false}, "Node supports BIP324 v2 transport protocol"},
```
done.

stratospher commented on src/rpc/net.cpp:
```diff
@@ -388,11 +389,12 @@ static RPCHelpMan addconnection()
     } else {
         throw JSONRPCError(RPC_INVALID_PARAMETER, self.ToString());
     }
+    bool use_v2transport = !request.params[2].isNull() && request.params[2].get_bool();
```
done.

stratospher commented on src/net.cpp:
```diff
@@ -1880,7 +1880,7 @@ void CConnman::CreateNodeFromAcceptedSocket(std::unique_ptr<Sock>&& sock,
 }
 
-bool CConnman::AddConnection(const std::string& address, ConnectionType conn_type)
+bool CConnman::AddConnection(const std::string& address, ConnectionType conn_type, bool use_v2transport = false)
```
done.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +97,143 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
             # Responding, place their public key encoding first.
             return TaggedHash("bip324_ellswift_xonly_ecdh", ellswift_theirs + ellswift_ours + ecdh_point_x32)
+
+    def initiate_v2_handshake(self, garbage_len=random.randrange(4096)):
```
oops, fixed.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +97,143 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+        return processed_length, False
+
+    def initialize_v2_transport(self, ecdh_secret):
+        """Return a peer object with various BIP324 derived keys and ciphers."""
```
done.

stratospher commented:
Rebased on master.

stratospher commented:
Updated since #28374 is merged.

theStack commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +100,147 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+        response = response[16:]
+        processed_length = len(received_garbage)
+        for i in range(4096):
+            if received_garbage[-16:] == self.peer['recv_garbage_terminator']:
+                # Receive, decode, and ignore version packet.
+                # This includes skipping decoys and authenticating the received garbage.
+                aad = received_garbage[:-16]
```
nit: as `received_garbage` is ensured to have a maximum size of 16 bytes a few lines above, the slicing isn't needed here.
```suggestion
            if received_garbage == self.peer['recv_garbage_terminator']:
                # Receive, decode, and ignore version packet.
                # This includes skipping decoys and authenticating the received garbage.
                aad = received_garbage
```

stratospher replied:
but there's also an else condition below where if the `received_garbage` doesn't match the garbage terminator, we keep appending 1 byte to `received_garbage` until there's a match with garbage terminator/we've exceed max garbage limit. so even though initially `received_garbage` length is 16 bytes, it can take values from 17 bytes, ..., 17 + 4095 bytes.

theStack replied:
Oh right, I missed the else-branch and assumed `received_garbage` wouldn't change in the loop (which wouldn't make much sense), nevermind.

theStack commented on test/functional/test_framework/test_node.py:
```diff
@@ -689,14 +700,34 @@ def add_outbound_p2p_connection(self, p2p_conn, *, wait_for_verack=True, p2p_idx
         p2p_conn.p2p_connected_to_node = False
-        p2p_conn.peer_accept_connection(connect_cb=addconnection_callback, connect_id=p2p_idx + 1, net=self.chain, timeout_factor=self.timeout_factor, **kwargs)()
+        if advertise_v2_p2p:
+            kwargs['services'] = kwargs['services']|NODE_P2P_V2 if 'services' in kwargs else P2P_SERVICES|NODE_P2P_V2
```
readability nit:
```suggestion
            kwargs['services'] = (kwargs['services']|NODE_P2P_V2) if 'services' in kwargs else (P2P_SERVICES|NODE_P2P_V2)
```

stratospher replied:
done.

theStack commented on test/functional/test_framework/p2p.py:
```diff
@@ -853,9 +855,11 @@ def send_blocks_and_test(self, blocks, node, *, success=True, force_send=False,
 
         reject_reason = [reject_reason] if reject_reason else []
         with node.assert_debug_log(expected_msgs=reject_reason):
+            if is_decoy:  # since decoy messages don't get sent - no need to wait for response
```
nit:
```suggestion
            if is_decoy:  # since decoy messages are ignored by the recipient - no need to wait for response
```
(or "don't get processed" or something like that?)

stratospher replied:
done.

theStack commented on test/functional/p2p_v2_encrypted.py:
```diff
@@ -0,0 +1,116 @@
+    check_node_connections,
+)
+
+REKEY_INTERVAL = 224
```
nit: could import this constant from test_framework.crypto

stratospher replied:
done.

theStack reviewed (commented):
Concept ACK

Did only a rough first review round so far, left some comments on the way, mostly nits. 

It seems like the newly introduced functional test `p2p_v2_encrypted.py` could be merged with the already existing `p2p_v2_transport.py` (here or in a follow-up), or is it fundamentally different?

mzumsande commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,10 +4,92 @@
 
+logger = logging.getLogger("TestFramework.v2_p2p")
+
+MAGIC_BYTES = {
+    "regtest": b"\xfa\xbf\xb5\xda"   # regtest
+}
```
I agree that would be nice, in particular so that we can enable `p2p_dos_header_tree.py` (which uses testnet instead of regtest) in combination with v2 transport. One possibility would be to move the `MAGIC_BYTES` constants from `p2p.py` into another file (e.g. `test_framework/messages.py`), change all includes, pass `net` to `EncryptedP2PState`, and change `V1_PREFIX` into a local variable using it)

stratospher replied:
nice! included `MAGIC_BYTES` in messages.py

mzumsande replied:
Thanks! 
I think you should actually use `net` by making all callers from `p2p.py`  pass it to `EncryptedP2PState`, the default value of 'regtest' should be dropped in my opinion. You can verify that it works when `p2p_dos_header_tree.py --transport` doesn't fail anymore.

Could also change all existing imports for MAGIC_BYTES (there are a few in other tests) to avoid indirect imports.

stratospher replied:
sounds much better! added a [new commit](https://github.com/bitcoin/bitcoin/pull/24748/commits/7f0e797dc9524c03f650f60a028cb46852877284).

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,10 +4,92 @@
+        - see section #overall_packet_encryption_and_decryption_pseudocode in BIP 324
+    """
+    def __init__(self, **kwargs):
+        self.initiating = kwargs['initiating']  # True if initiator
```
done.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +100,147 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+        """Initiator begins the v2 handshake by sending its ellswift bytes and garbage"""
+        self.privkey_ours, self.ellswift_ours = ellswift_create()
+        garbage_len = random.randrange(4096)
+        self.sent_garbage = os.urandom(garbage_len)
```
true! done.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +100,147 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+            if self.received_prefix[-1] != V1_PREFIX[len(self.received_prefix) - 1]:
+                self.privkey_ours, self.ellswift_ours = ellswift_create()
+                garbage_len = random.randrange(4096)
+                self.sent_garbage = os.urandom(garbage_len)
```
makes sense. done.

stratospher commented:
@theStack, @mzumsande thank you for the reviews! I've rebased and addressed your comments.

> It seems like the newly introduced functional test p2p_v2_encrypted.py could be merged with the already existing p2p_v2_transport.py (here or in a follow-up), or is it fundamentally different?

yes! can be done in a follow up. only difference is `p2p_v2_encrypted.py` tests `TestNode` <--> `P2PInterface` behaviour and `p2p_v2_transport.py` tests `TestNode` <--> `TestNode` behaviour. 

@maflcko, still need to address the CI failure. https://api.cirrus-ci.com/v1/task/6171895557521408/logs/ci.log

theStack reviewed (commented):
I've noticed that the test `p2p_v2_encrypted.py` fails if the node under test sends a small amount of garbage (in the range of 0-19 bytes). Can be reproduced by applying the following simple patch:
```diff
diff --git a/src/net.cpp b/src/net.cpp
index 2682035912..09af15a9b7 100644
--- a/src/net.cpp
+++ b/src/net.cpp
@@ -951,7 +951,7 @@ std::vector<uint8_t> GenerateRandomGarbage() noexcept
 {
     std::vector<uint8_t> ret;
     FastRandomContext rng;
-    ret.resize(rng.randrange(V2Transport::MAX_GARBAGE_LEN + 1));
+    ret.resize(19); // fails in range 0-19, works with 20+
     rng.fillrand(MakeWritableByteSpan(ret));
     return ret;
 }
```
Apparently the expected VERACK is not received from the node. Didn't dig deeper yet. For finding the cause, I think some `logger.debug(...)` messages in `v2_p2p.py` might be helpful.

stratospher committed:
[test] Move MAGIC_BYTES to messages.py

This avoids circular dependency happening when importing MAGIC_BYTES.
Before,
	p2p.py <--import for EncryptedP2PState-- v2_p2p.py
	  |					    ^
	  |				            |
	  └---------import for MAGIC_BYTES----------┘
Now, MAGIC_BYTES are kept separately in messages.py

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>

stratospher commented:
> I've noticed that the test p2p_v2_encrypted.py fails if the node under test sends a small amount of garbage (in the range of 0-19 bytes).

thanks for catching that @theStack! before the drop garbage authentication packet change in #28525, `recvbuf` needed 16+20+20 bytes to authenticate the handshake. now it only needs 16+20 bytes. i've fixed it now.

stratospher commented:
>  https://api.cirrus-ci.com/v1/task/6171895557521408/logs/ci.log

i wasn't able to reproduce this error - tried it on the docker container, setting randomseed and running it in loop. also confused how it could have happened.

this is what happens inside `p2p_v2_earlykeyresponse.py` test
- we start an inbound connection to the TestNode (`TestNode` <---- `P2PInterface`). 
- v2 connection starts by sending 64 bytes ellswift (`TestNode` <--64 bytes-- `P2PInterface`). this is sent in 2 parts:
	- `TestNode` <--4 bytes magic bytes-- `P2PInterface` - `TestNode` doesn't send us a response
	- `TestNode` <--remaining 60 bytes-- `P2PInterface` - `TestNode` responds back since there's a mismatch from V1_PREFIX. 

What usually happens:
1. During inbound connection creation, `TestNode::add_p2p_connection()` calls `P2PConnection::peer_connect()` where an object of `EncryptedP2PState` class is created - this object `v2_state` contains the function `EncryptedP2PState::initiate_v2_handshake()` to send ellswift bytes
2. then the asyncio code to create a connection happens and `connection_made()` is called when the connection is opened
	- inside `connection_made()`, 64 bytes ellswift is sent using `EncryptedP2PState::initiate_v2_handshake()`
3. When the `TestNode` sends a response, `P2PConnection::data_received()` callback will detect it


What happens in test:
- there are 2 variables:
1. `send_net_magic` - initially true. after we send magic bytes, `send_net_magic_bytes` is set to false (basically just a switch to send 64 bytes in 2 parts)
2. `can_data_be_received` - initally false - data can't be received until mismatch from V1_PREFIX occurs. after we send remaining mismatched 60 bytes, `can_data_be_received` is set to true.
- since response will be detected on `TestNode`, we add an assertion to make sure `send_net_magic` = false (meaning magic bytes already sent) and `can_data_be_received` = true (meaning mismatched 60 bytes also sent). `TestNode` wouldn't send a response in the intermediate states.
- so the possible states are:
1. initially
	`send_net_magic` = true, `can_data_be_received` = false (Assertion would fail in `data_received()` but shouldn't happen)
2. after `connection_made()`'s `custom EncryptedP2PState::initiate_v2_handshake()` (TestNode <--4 bytes magic bytes-- P2PInterface)
	`send_net_magic` = false, `can_data_be_received` = false (Assertion would fail in `data_received()` but shouldn't happen)
3. after `p2p_v2_earlykeyresponse.py` test when we call `custom EncryptedP2PState::initiate_v2_handshake()` explicitly (TestNode <--remaining 60 bytes-- P2PInterface)
	`send_net_magic` = false, `can_data_be_received` = true (Assertion passes in `data_received()`)

mzumsande commented on test/functional/p2p_v2_earlykeyresponse.py:
```diff
@@ -0,0 +1,86 @@
+        """
+        if self.send_net_magic:
+            self.privkey_ours, self.ellswift_ours = ellswift_create()
+            self.sent_garbage = os.urandom(garbage_len)
```
one more instance of os.urandom (`random.randbytes(garbage_len)` should work)

stratospher replied:
done.

mzumsande reviewed (commented):
I had a look at `p2p_v2_earlykeyresponse.py` and it would fail the handshake rarely (but the test still passes because the error is thrown in the p2p thread!).
I could reproduce it by changing `GenerateRandomGarbage()` similar to TheStack above, but this time to large values:
`ret.resize(V2Transport::MAX_GARBAGE_LEN-19);` or larger. I haven't analyzed any further so far.

[Edit]: Well, it is expected that the handshake fails, because we manipulated our key we'll never find the expected garbage terminator. But it seems confusing to fail with a stack trace / uncaught exception in this case ("`Fatal error: protocol.data_received() call failed.`") when this is actually expected to happen.

mzumsande commented:
Ah, I think if we have an unhandled exception like in my previous post, running the test directly (`python3 p2p_v2_earlykeyresponse.py`) will show success, but if we use the test_runner (`python3 test_runner.py p2p_v2_earlykeyresponse.py`) it will show failure, probably because things were written to stderr. So we need to do something about the case I mentioned above.

stratospher commented:
> But it seems confusing to fail with a stack trace / uncaught exception in this case ("Fatal error: protocol.data_received() call failed.") when this is actually expected to happen.


<details>
<summary> is this the test log you observed? </summary>
<br>

```
2023-11-18T13:52:32.976000Z TestFramework (INFO): PRNG seed is: 8699047777059301467
2023-11-18T13:52:32.977000Z TestFramework (INFO): Initializing test directory /tmp/bitcoin_func_test_70zlmcgr
2023-11-18T13:52:33.535000Z TestFramework (INFO): Sending ellswift bytes in parts to ensure that response from responder is received only when
2023-11-18T13:52:33.536000Z TestFramework (INFO): ellswift bytes have a mismatch from the 16 bytes(network magic followed by "version\x00\x00\x00\x00\x00")
2023-11-18T13:52:33.536000Z TestFramework (INFO): Sending first 4 bytes of ellswift which match network magic
2023-11-18T13:52:33.536000Z TestFramework (INFO): If a response is received, assertion failure would happen in our custom data_received() function
2023-11-18T13:52:33.587000Z TestFramework (INFO): Sending remaining ellswift and garbage which are different from V1_PREFIX. Since a response is
2023-11-18T13:52:33.587000Z TestFramework (INFO): expected now, our custom data_received() function wouldn't result in assertion failure
Fatal error: protocol.data_received() call failed.
protocol: <__main__.PeerEarlyKey object at 0x7f0d6f6c1060>
transport: <_SelectorSocketTransport fd=10 read=polling write=<idle, bufsize=0>>
Traceback (most recent call last):
  File "/usr/lib/python3.10/asyncio/selector_events.py", line 876, in _read_ready__data_received
    self._protocol.data_received(data)
  File "/home/stratospher/code/bitcoin/test/functional/p2p_v2_earlykeyresponse.py", line 60, in data_received
    super(PeerEarlyKey, self).data_received(t)
  File "/home/stratospher/code/bitcoin/test/functional/test_framework/p2p.py", line 297, in data_received
    self.v2_handshake()
  File "/home/stratospher/code/bitcoin/test/functional/test_framework/p2p.py", line 284, in v2_handshake
    raise ValueError("invalid v2 mac tag in handshake authentication")
ValueError: invalid v2 mac tag in handshake authentication
2023-11-18T13:52:33.673000Z TestFramework.p2p (WARNING): Connection lost to 127.0.0.1:13882 due to invalid v2 mac tag in handshake authentication
2023-11-18T13:52:33.693000Z TestFramework (INFO): successful disconnection when MITM happens in the key exchange phase
2023-11-18T13:52:33.744000Z TestFramework (INFO): Stopping nodes
2023-11-18T13:52:33.948000Z TestFramework (INFO): Cleaning up /tmp/bitcoin_func_test_70zlmcgr on exit
2023-11-18T13:52:33.948000Z TestFramework (INFO): Tests successful
```
</details>


this happened because of the mac tag mismatch in the v2 handshake's authentication. it only got displayed sometimes because the test would finish before the v2 handshake authentication happened.  since we don't need to do v2 handshake authentication/v2 handshake for this test, and just be able to detect when data is received, i've removed the `P2PInterface::data_received()` call which performs the handshake. whenever data is received, `PeerEarlyKey::data_received()` would detect it and do nothing with the received bytes.


> I could reproduce it by changing GenerateRandomGarbage() similar to TheStack above, but this time to large values:
ret.resize(V2Transport::MAX_GARBAGE_LEN-19); or larger.

oh! i'm not able to reproduce this by changing to large values. Could this be related to the above test log?

stratospher commented on test/functional/p2p_v2_earlykeyresponse.py:
```diff
@@ -0,0 +1,86 @@
+    def connection_made(self, *args, **kwargs):
+        """Custom implementation so that 64 bytes ellswift is sent in 2 parts during `initial_v2_handshake()`"""
+        self.v2_state = TestEncryptedP2PState()
+        super(PeerEarlyKey, self).connection_made(*args, **kwargs)
```
you're right! done.

mzumsande commented:
> is this the test log you observed?

Yes. Only in ~1% of runs with the previous version, but after changing the garbage size in `net.cpp` to the max I would get it consistently.
I think the reason is that usually bitcoind would terminate the connection (because the python side would wait for more bytes that wouldn't come), but when the garbage was close to the max, the python side would terminate it and throwing the exception.

> i've removed the P2PInterface::data_received() call which performs the handshake

Thanks, that should fix this issue.

mzumsande commented on test/functional/p2p_v2_encrypted.py:
```diff
@@ -0,0 +1,117 @@
+                # check if node1 connected to node0 (but not to node0's p2p connection directly)
+                # gets blocks produced by node0's p2p connection
+                self.log.info("Check if blocks produced by node0's p2p connection is received by node0")
+                peer6.send_blocks_and_test(test_blocks, node0, success=True) # node0's tip advances
```
One possible problem I encountered when running the `p2p_invalid_messages.py` `test_resource_exhaustion` subtest with v2 that but could affect many tests intermittently, also this spot:

I think that there could be a race between the python main thread of functional tests and the p2p thread:
For example, this command 1) builds messages, 2) encrypts them and 3) sends them, all within the main thread.
If during step 2) the bitcoind node would decide to send out a ping, the p2p thread will also encrypt a message (the pong answer) and send it. It doesn't wait for the other thread. This isn't a problem with v1 (if the messages are sent in different orders, nothing may break),  but with v2 I think the connection would break down because the ChaCha stream cipher is stateful.

So I think that maybe we need some kind of synchronization / locking mechanism?

I haven't though much about a solution yet, just wanted to share this problem.

theStack replied:
Good point indeed. To execute the steps build/encrypt/send-message atomically, I think introducing a new lock and acquiring it in the `send_message` method should do the trick? I.e.
```diff
diff --git a/test/functional/test_framework/p2p.py b/test/functional/test_framework/p2p.py
index a70682a768..59707471a6 100755
--- a/test/functional/test_framework/p2p.py
+++ b/test/functional/test_framework/p2p.py
@@ -164,6 +164,7 @@ class P2PConnection(asyncio.Protocol):
         # The underlying transport of the connection.
         # Should only call methods on this from the NetworkThread, c.f. call_soon_threadsafe
         self._transport = None
+        self._send_lock = threading.Lock()
         self.v2_state = None  # EncryptedP2PState object needed for v2 p2p connections
         self.supports_v2_p2p = False  # set if the connection supports v2 p2p
         self.reconnect = False  # set if reconnection needs to happen
@@ -371,9 +372,10 @@ class P2PConnection(asyncio.Protocol):
 
         This method takes a P2P payload, builds the P2P header and adds
         the message to the send buffer to be sent over the socket."""
-        tmsg = self.build_message(message, is_decoy)
-        self._log_message("send", message)
-        return self.send_raw_message(tmsg)
+        with self._send_lock:
+            tmsg = self.build_message(message, is_decoy)
+            self._log_message("send", message)
+            return self.send_raw_message(tmsg)
 
     def send_raw_message(self, raw_message_bytes):
         if not self.is_connected:
```
It might be also good to classify the methods `build_message` and `send_raw_message` as private then (that is, prefixing them with an underscore) to express that they shouldn't be called from the outside anymore, as this would cause problems for v2 connections after handshake completion. Currently only `p2p_invalid_messages.py` is calling those methods, so that test needs to be reworked for v2 support (didn't look deeper yet though, can maybe also done in a follow-up).

stratospher replied:
+1, good catch! when i tried this, just running `self.test_resource_exhaustion()` in `p2p_invalid messages.py` took too much time. 
1. in v1, without locks - 0m22.265s
2. in v2, without locks - won't work
3. in v1, with locks - 0m22.101s
4. in v2, with locks - 8m45.916s

theStack replied:
@stratospher: Seems like the long run-time in case 4. is primarily caused by the slow ChaCha20 implementation in Python. On my machine, encrypting 4MB of data (as done in `test_resource_exhaustion`, but repeated 80 times) takes already more than half a minute:
```
$ ipython 
Python 3.10.13 (main, Oct  5 2023, 16:21:31) [Clang 13.0.0 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.13.2 -- An enhanced Interactive Python. Type '?' for help.

In [1]: from test.functional.test_framework.crypto.bip324_cipher import aead_chacha20_poly1305_encrypt

In [2]: %time _ = aead_chacha20_poly1305_encrypt(key=bytes([1]*32), nonce=bytes([2]*12), aad=bytes(), plaintext=bytes([0xcc]*4000000))
CPU times: user 35.9 s, sys: 40 ms, total: 36 s
Wall time: 35.8 s
```
I guess we can't do that much about that (unless someone sees a way to optimize our python cryptography, without introducing new dependencies like `numpy` or calling an external library?), other than reducing the number of loops in `test_resource_exhaustion` if p2p-v2 is used.

mzumsande replied:
> so that test needs to be reworked for v2 support (didn't look deeper yet though, can maybe also done in a follow-up)

Yes, multiple tests need an adjustment for running with python v2 P2P. I have a WIP branch at 
https://github.com/mzumsande/bitcoin/tree/tmp_bip324_fixalltests for that which has helped me find this concurrency issue and other issues I mentioned above, but will only open a PR after this PR is merged.

> Seems like the long run-time in case 4. is primarily caused by the slow ChaCha20 implementation in Python.

I agree. I think that his particular subtest maybe could be skipped with v2 because its goal is to stress the node by bombarding it with lots of messages in a short time, not the test framework. But also,  `feature_block.py` and `feature_maxuploadtarget.py`, become very slow on my computer for the same reason and we might have to only run them with v1 transport.

stratospher replied:
> I guess we can't do that much about that (unless someone sees a way to optimize our python cryptography, without introducing new dependencies like numpy or calling an external library?), other than reducing the number of loops in test_resource_exhaustion if p2p-v2 is used.

makes sense.

i've introduced a new lock in [this commit](https://github.com/bitcoin/bitcoin/pull/24748/commits/99f162f3abfe5f76f914a20169743cd151d6c14e).

stratospher committed:
[test] Use lock for sending P2P messages in test framework

Messages are built, encrypted and sent over the socket in v2
connections. If a race condition happens between python's main
thread and p2p thread with both of them trying to send a message,
it's possible that the messages get encrypted with wrong keystream.

Messages are built and sent over the socket in v1 connections.
So there's no problem if messages are sent in the wrong order.

Co-authored-by: Martin Zumsande <mzumsande@gmail.com>
Co-authored-by: theStack <sebastian.falbesoner@gmail.com>

mzumsande reviewed (commented):
Another round of review - looking good!

stratospher commented on test/functional/test_framework/test_node.py:
```diff
@@ -197,6 +198,10 @@ def start(self, extra_args=None, *, cwd=None, stdout=None, stderr=None, env=None
         if extra_args is None:
             extra_args = self.extra_args
+        if "-v2transport=1" in extra_args:
+            self.advertise_v2_p2p = True
```
yay! done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -231,12 +231,65 @@ def connection_lost(self, exc):
+            # `respond_v2_handshake()`.
+            # `complete_handshake()` reads the remaining `64 - length` ellswift bytes afterwards from recvbuf.
+            # if we're the initiator, length = 0 and 64 bytes ellswift is read from recvbuf in `complete_handshake()`
+            if len(self.recvbuf) < 64 - length:
```
good catch! done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -231,12 +231,65 @@ def connection_lost(self, exc):
+        # `self.v2_state.peer` is instantiated only after shared ECDH secret/BIP324 derived keys and ciphers
+        # is derived in `complete_handshake()`.
+        # so `authenticate_handshake()` which uses the BIP324 derived ciphers gets called after `complete_handshake()`.
+        if self.v2_state.peer:
```
you're right! i've made it an assert.

i vaguely remember adding this in some older version of the branch because when `TestNode` sends transport layer packets(64 bytes ellswift, x bytes garbage, 16 bytes garbage terminator, optional decoy packets, 20 bytes transport version packet) to our python `P2PInterface`, it's not necessary that we receive all these bytes at once on the wire/have access to these bytes at once on `self.recvbuf`. Looking at this again - even without the "`if self.v2_state.peer`", the code paths for when 64 + 16 + 20 bytes is received or when 64 bytes, delay, 16 + 20 bytes is received works fine!

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -231,12 +231,65 @@ def connection_lost(self, exc):
+        # so `authenticate_handshake()` which uses the BIP324 derived ciphers gets called after `complete_handshake()`.
+        if self.v2_state.peer:
+            # authenticate v2 handshake
+            if len(self.recvbuf) < 16 + 20:
```
added!

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -553,6 +560,11 @@ def on_verack(self, message):
+        if self.reconnect:
+            while self.queue_messages:
+                message = self.queue_messages.pop(0)
+                self.send_message(message)
```
good point! done.

stratospher commented:
thanks @mzumsande for the really great feedback! I’ve rebased on master since #28805 is merged and updated the PR to address your comments.

theStack commented on test/functional/p2p_v2_encrypted.py:
```diff
@@ -0,0 +1,117 @@
+
+    def run_test(self):
+        node0, node1 = self.nodes[0], self.nodes[1]
+        self.log.info("Check inbound connections to v2 TestNode from v2 P2PConnection is v2")
+        peer1 = node0.add_p2p_connection(P2PInterface(), wait_for_verack=True, supports_v2_p2p=True)
+        assert peer1.supports_v2_p2p
```
to verify each connection's transport protocol version, could additionally take the node's perspective into account via `getpeerinfo()`, e.g.:
<details>
<summary>diff</summary>

```diff
diff --git a/test/functional/p2p_v2_encrypted.py b/test/functional/p2p_v2_encrypted.py
index 177c3ed42f..8f5c4bb925 100755
--- a/test/functional/p2p_v2_encrypted.py
+++ b/test/functional/p2p_v2_encrypted.py
@@ -56,23 +56,28 @@ class P2PEncrypted(BitcoinTestFramework):
         self.log.info("Check inbound connections to v2 TestNode from v2 P2PConnection is v2")
         peer1 = node0.add_p2p_connection(P2PInterface(), wait_for_verack=True, supports_v2_p2p=True)
         assert peer1.supports_v2_p2p
+        assert_equal(node0.getpeerinfo()[-1]["transport_protocol_type"], "v2")
 
         self.log.info("Check inbound connection to v2 TestNode from v1 P2PConnection is v1")
         peer2 = node0.add_p2p_connection(P2PInterface(), wait_for_verack=True, supports_v2_p2p=False)
         assert not peer2.supports_v2_p2p
+        assert_equal(node0.getpeerinfo()[-1]["transport_protocol_type"], "v1")
 
         self.log.info("Check outbound connection from v2 TestNode to v1 P2PConnection advertised as v1 is v1")
         peer3 = node0.add_outbound_p2p_connection(P2PInterface(), p2p_idx=0, supports_v2_p2p=False, advertise_v2_p2p=False)
         assert not peer3.supports_v2_p2p
+        assert_equal(node0.getpeerinfo()[-1]["transport_protocol_type"], "v1")
 
         # v2 TestNode performs downgrading here
         self.log.info("Check outbound connection from v2 TestNode to v1 P2PConnection advertised as v2 is v1")
         peer4 = node0.add_outbound_p2p_connection(P2PInterface(), p2p_idx=1, supports_v2_p2p=False, advertise_v2_p2p=True)
         assert not peer4.supports_v2_p2p
+        assert_equal(node0.getpeerinfo()[-1]["transport_protocol_type"], "v1")
 
         self.log.info("Check outbound connection from v2 TestNode to v2 P2PConnection advertised as v2 is v2")
         peer5 = node0.add_outbound_p2p_connection(P2PInterface(), p2p_idx=2, supports_v2_p2p=True, advertise_v2_p2p=True)
         assert peer5.supports_v2_p2p
+        assert_equal(node0.getpeerinfo()[-1]["transport_protocol_type"], "v2")
 
         self.log.info("Check if version is sent and verack is received in inbound/outbound connections")
         assert_equal(len(node0.getpeerinfo()), 5)  # check if above 5 connections are present in node0's getpeerinfo()
@@ -88,6 +93,7 @@ class P2PEncrypted(BitcoinTestFramework):
             # Add v2 P2P connection to node0
             peer6 = node0.add_p2p_connection(P2PDataStore(), supports_v2_p2p=True)
             assert peer6.supports_v2_p2p
+            assert_equal(node0.getpeerinfo()[-1]["transport_protocol_type"], "v2")
 
             if i:
                 # check if node1 connected to node0 (but not to node0's p2p connection directly)
@@ -115,6 +121,7 @@ class P2PEncrypted(BitcoinTestFramework):
         self.restart_node(0, ["-v2transport=0"])
         peer7 = node0.add_p2p_connection(P2PInterface(), wait_for_verack=True, supports_v2_p2p=True)
         assert not peer7.supports_v2_p2p
+        assert_equal(node0.getpeerinfo()[-1]["transport_protocol_type"], "v1")
         check_node_connections(node=node0, num_in=1, num_out=0)
```
</details>
(same for the downgrading check commit after)

stratospher replied:
good suggestion! done.

theStack reviewed (commented):
Left some comments below from the latest review round, mostly nits. I think another small check to add in the newly introduced functional test could be to verify that a `P2PConnection`'s session id matches the one from the node, as returned via `getpeerinfo()`.

mzumsande commented on test/functional/test_framework/test_node.py:
```diff
@@ -642,18 +643,30 @@ def assert_start_raises_init_error(self, extra_args=None, expected_msg=None, mat
         p2p_conn.p2p_connected_to_node = True
-        p2p_conn.peer_connect(**kwargs, send_version=send_version, net=self.chain, timeout_factor=self.timeout_factor)()
+        if self.use_v2transport:
+            kwargs['services'] = kwargs['services']|NODE_P2P_V2 if 'services' in kwargs else P2P_SERVICES|NODE_P2P_V2
```
Ah, now that `use_v2transport` with the correct restart behavior is used, CI fails because it is necessary to fix the `test_service_flags` subtest from rpc_net.py already in this PR, to something like
```
if self.options.v2transport:
    assert_equal(['UNKNOWN[2^4]', 'P2P_V2', 'UNKNOWN[2^63]'], self.nodes[0].getpeerinfo()[-1]['servicesnames'])
else:
    assert_equal(['UNKNOWN[2^4]', 'UNKNOWN[2^63]'], self.nodes[0].getpeerinfo()[-1]['servicesnames'])
```

stratospher replied:
missed that! i've updated the subtest in `rpc_net.py`.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,10 +4,90 @@
+def GetShortIDFromMessageType(msgtype):
+    """Returns 1-byte short message type ID for the P2P message"""
+    msgtype_to_shortid = dict(map(reversed, SHORTID.items()))
+    return msgtype_to_shortid[msgtype].to_bytes(1, 'big') if msgtype in msgtype_to_shortid else b"\x00"
```
interesting! i've kept a `MSGTYPE_TO_SHORTID` dictionary and used this construct in [`build_message`](https://github.com/bitcoin/bitcoin/blob/6567fe2aafa43b99e8298018b0665222eec4daff/test/functional/test_framework/p2p.py#L398).

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +98,148 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+        self.privkey_ours, self.ellswift_ours = ellswift_create()
+        garbage_len = random.randrange(MAX_GARBAGE_LEN + 1)
+        self.sent_garbage = random.randbytes(garbage_len)
+        logger.debug("sending %d bytes of garbage data" % garbage_len)
```
done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -387,17 +388,29 @@ def maybe_write():
-        return tmsg
+        if self.supports_v2_p2p:
+            msgtype = message.msgtype
+            data = message.serialize()
```
done.

stratospher commented:
thanks @theStack, @mzumsande! I've updated the PR to address your comments.
Also made this change to `MSGTYPE_TO_SHORTID` construction because of `lint-python.py` failure in the CI

```diff
diff --git a/test/functional/test_framework/v2_p2p.py b/test/functional/test_framework/v2_p2p.py
index fe44351a13..b4efd1ac58 100644
--- a/test/functional/test_framework/v2_p2p.py
+++ b/test/functional/test_framework/v2_p2p.py
@@ -55,7 +55,7 @@ SHORTID = {
 }
 
 # Dictionary which contains short message type ID for the P2P message
-MSGTYPE_TO_SHORTID = dict(map(reversed, SHORTID.items()))
+MSGTYPE_TO_SHORTID = {msgtype: shortid for shortid, msgtype in SHORTID.items()}
 
 
 class EncryptedP2PState:
```

mzumsande commented on test/functional/test_framework/p2p.py:
```diff
@@ -163,6 +163,7 @@ def __init__(self):
         # The underlying transport of the connection.
         # Should only call methods on this from the NetworkThread, c.f. call_soon_threadsafe
         self._transport = None
+        self._send_lock = threading.Lock()
```
One thing to consider is that now we have two locks for p2p, we need to worry about deadlocks.
There is an implied lock order, `on_message` locks `p2p_lock`, and can call messages (e.g. `on_inv`) that will call `send_message` and lock `_send_lock`. So we must make sure never to take `p2p_lock` after `_send_lock`. I don't think that this is currently possible, but maybe we should add a comment about this.

stratospher replied:
good find! i've added a comment.

stratospher replied:
@mzumsande, @sr-gi, I played with @sr-gi's idea on [this branch](https://github.com/stratospher/bitcoin/commit/276b95d766cd11d65d563805d52beace8dd32744) and wondering about 2 things:

1. can deadlocks happen? i noticed places in the test where `_send_lock` was acquired and then `p2p_lock` (and deadlocks didn't happen). am i missing something?
	1. MainThread acquires `_send_lock`
	2. NetworkThread acquires `p2p_lock`
	3. NetworkThread would wait for MainThread to release `_send_lock`
	4. MainThread would release `_send_lock` at some point of time right? (since it doesn't need to wait for `p2p_lock`)
2. do we need a separate `_send_lock` for every P2PConnection? benchmarks look slightly faster with a global `_send_lock` (like `p2p_lock`).

> Given this will affect several parts of the tests suite it may be worth doing it as a followup

+1, but supposing deadlocks are possible - in the follow up:
1. `_send_lock` acquired because of `send_message()` and then `p2p_lock` acquired because of `wait_until()` is harmless for us
2. `_send_lock` acquired because of `send_message()` and then `p2p_lock` acquired because of `on_message()` is the situation we need to be concerned about

sr-gi replied:
> do we need a separate `_send_lock` for every P2PConnection? benchmarks look slightly faster with a global `_send_lock` (like `p2p_lock`).

Umm, that sounds a bit strange to me. My intuition would be that having a global lock should be slower, especially if multiple connections are run at the same time, given they would be racing to acquire that lock every time a message needs to be sent (?)

w.r.t wether or not a deadlock is possible, I think @mzumsande may be able to tell you better given he was the one pointing them out initially, but if we want to enforce a certain ordering of acquiring them I think my suggestion (or something on those lines) may be worth adding. I'm even happy to open a PR for that

mzumsande replied:
> Something along the lines of:
> 
> ```python
> def p2p_lock(self):
>     assert not self._send_lock.locked()
>     return p2p_lock
> ```

I think that this is not right (I tried it out, and it asserts everywhere :sweat_smile:):
To prevent deadlocks, the lock order must be preserved only within each thread:
- It's possible that a given thread first takes `p2p_lock`, and then `send_lock`. See for example: `on_message` -> `p2p_lock` taken -> call `on_inv` -> call `send_message` -> `_send_lock` taken
- What must be disallowed is that any thread can do the reverse: First take `_send_lock`, and then take `p2p_lock`, because it would lead to a deadlockif one thread did that, and another thread did the reverse. 
- I don't think that this is possible because I looked at the one spot where `send_lock` is taken (it calls `self.build_message` and `self.send_raw_message`, neither of which take `p2p_lock`) 
- But it is perfectly allowed that Thread 1 take `p2p_lock`, and immediately after Thread 2 takes `_send_lock` (which would trigger the assert).

sr-gi replied:
You're right. I don't think there is any easy way of keeping track of who's thread is holding what lock in Python though, so I don't think there a straightforward way of addressing this :(

mzumsande reviewed (commented):
ACK f479a99cc92c7966266804e8c33338cebbc44fbc

I have reviewed the code. I did some testing / gained more confidence that it works well by having a [branch](https://github.com/mzumsande/bitcoin/tree/tmp_bip324_fixalltests) where, with some adjustments, the python v2 p2p module is used for the entire functional test suite.

sipa reviewed (commented):
Halfway through.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,10 +4,87 @@
+
 
 class EncryptedP2PState:
+    """A class for performing v2 P2P protocol functions on P2PConnection:
```
definitely useful to have. done.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +95,148 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+        msg_to_send = self.peer['send_garbage_terminator']
+        # Optionally send decoy packets after garbage terminator.
+        aad = self.sent_garbage
+        for decoy_content_len in [random.randint(1, 100) for _ in range(10)]:
```
changed it to send 0-9 number of decoy packets randomly.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +95,148 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+                      - return 0 if only part of packet is received. (recvbuf not updated since decryption not done yet)
+                      - return -1 if there's a MAC tag mismatch and disconnect.
+            2. decrypted packet contents
+                     - return b"" if only part of packet is received/MAC tag mismatch.
```
good catch! changed it to `None`.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +95,148 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+                        return processed_length, False
+                    elif length == 0:
+                        return 0, True
+                    # currently version packet has empty content while decoy packets don't
```
right! done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -199,8 +211,13 @@ def connection_made(self, transport):
+        # in an inbound connection to the TestNode, P2PConnection is the initiator. [TestNode <---- P2PConnection]
+        # ellswift is generated and sent immediately to begin the initial v2 handshake.
+        if self.supports_v2_p2p and self.v2_state.initiating and not self.v2_state.tried_v2_handshake:
+            ellswift_and_garbage_data = self.v2_state.initiate_v2_handshake()
```
oh makes sense to `s/ellswift_and_garbage_data/send_handshake_bytes` in `p2p.py`. didn't change the name in a test where we [send ellswift in parts](https://github.com/bitcoin/bitcoin/blob/ad0ae3d2128d04ff4f62a4bf612286d153dc314b/test/functional/p2p_v2_earlykeyresponse.py#L78).

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +95,148 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+                aad = received_garbage[:-16]
+                while not self.tried_v2_handshake:
+                    length, contents = self.v2_receive_packet(response, aad=aad)
+                    processed_length += length
```
done.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -18,3 +95,148 @@ def v2_ecdh(priv, ellswift_theirs, ellswift_ours, initiating):
+            if not byte:
+                return b""
+            self.received_prefix += byte
+            v1_prefix = MAGIC_BYTES[self.net] + b'version\x00\x00\x00\x00\x00'
```
nice, i've kept it outside the loop since we don't need it anywhere else for now.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -301,23 +302,44 @@ def _on_data(self):
+                    if shortid == 0:
+                        # a string command
+                        msgtype = msg[1:13].rstrip(b'\x00')
+                        msg = msg[13:]  # msg is set to be payload
```
done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -301,23 +302,44 @@ def _on_data(self):
+                        return
+                    # v2 P2P messages are read
+                    if len(self.recvbuf) < 3 + 1 + 16:
+                        return
```
true, replaced it with a check using `v2_receive_packet` return values which are anyways computed.

stratospher commented:
thanks for the reviews @mzumsande, @sipa, @theStack! i've updated the PR to address your comments. sorry for the delay since i was away.

theStack reviewed (approved):
ACK ad0ae3d2128d04ff4f62a4bf612286d153dc314b

stratospher commented on test/functional/test_framework/test_node.py:
```diff
@@ -642,18 +643,30 @@ def assert_start_raises_init_error(self, extra_args=None, expected_msg=None, mat
         p2p_conn.p2p_connected_to_node = True
-        p2p_conn.peer_connect(**kwargs, send_version=send_version, net=self.chain, timeout_factor=self.timeout_factor)()
+        if self.use_v2transport:
+            kwargs['services'] = kwargs['services']|NODE_P2P_V2 if 'services' in kwargs else P2P_SERVICES|NODE_P2P_V2
```
nice! done.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,17 +4,275 @@
+               - return 0 if only part of packet is received. (recvbuf not updated since decryption not done yet)
+               - return -1 if there's a MAC tag mismatch and disconnect.
+        2. bytes - decrypted packet contents
+                 - return b"" if only part of packet is received/MAC tag mismatch.
```
added a [description](https://github.com/bitcoin/bitcoin/blob/42203f84bba7e4ba206c9f951c3f4b6922a44f81/test/functional/test_framework/v2_p2p.py#L260-L265) and also an [extra assert](https://github.com/bitcoin/bitcoin/blob/42203f84bba7e4ba206c9f951c3f4b6922a44f81/test/functional/test_framework/p2p.py#L326) to make sure we don't receive b"" as an application layer message.

mzumsande commented on test/functional/test_framework/test_node.py:
```diff
@@ -741,7 +741,10 @@ def addconnection_callback(address, port):
         else:
-            p2p_conn.wait_for_connect()
+            if reconnect:
+                p2p_conn.wait_for_reconnect()
```
I think that this should be moved outside of the `if connection_type == "feeler":` clause and be done before:
After all the v2 reconnection logically stuff comes first, and then all the existing v1 logic still applies to the downgraded v1 connection - for example, the reconnection logic should also work if `add_outbound_p2p_connection` specified a feeler connection - in which case we currently wouldn't call `wait_for_reconnect`.

stratospher replied:
true! done.

mzumsande commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,17 +4,280 @@
+                    if contents is not None:
+                        self.tried_v2_handshake = True
+                        return processed_length, True
+                    response = response[length:]
```
just want to add that if this turned out annoying to fix (I didn't see an immediate trivial fix but haven't looked hard), I'd be fine just throwing an assert / error if decoy packages are received in this PR (since bitcoind shouldn't send those anyway).

stratospher replied:
great catch! i've changed the `authenticate_handshake` function to first deal with garbage terminator detection before processing decoy packets and version packet. hopefully this is not a concern now?

sr-gi commented on test/functional/p2p_v2_encrypted.py:
```diff
@@ -0,0 +1,123 @@
+
+    def run_test(self):
+        node0, node1 = self.nodes[0], self.nodes[1]
+        self.log.info("Check inbound connections to v2 TestNode from v2 P2PConnection is v2")
```
nit: connection (for consistency with the rest)

stratospher replied:
done.

sr-gi commented on test/functional/p2p_v2_encrypted.py:
```diff
@@ -0,0 +1,123 @@
+        self.log.info("Check the connections opened as expected")
+        check_node_connections(node=node0, num_in=4, num_out=2)
+
+        self.log.info("Check inbound connections to v1 TestNode from v2 P2PConnection is v1")
```
nit: connection (for consistency with the rest).

Also, it may be worth moving this right `peer2` so inbounds and outbounds are grouped

stratospher replied:
done.

> Also, it may be worth moving this right peer2 so inbounds and outbounds are grouped

didn't follow. i've renamed `peer7` to `peer1` since this is first peer in restarted node. is that what you meant?

stratospher replied:
oh that way! the last test was for v1 behaviour and peer7 needed a v1 `TestNode` while all the others needed a v2 `TestNode` which is why it was kept separately in the end.

sr-gi reviewed (commented):
First pass, untested, will do a more thorough review next week

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -174,16 +180,22 @@ def peer_connect_helper(self, dstaddr, dstport, net, timeout_factor):
         self.magic_bytes = MAGIC_BYTES[net]
 
-    def peer_connect(self, dstaddr, dstport, *, net, timeout_factor):
+    def peer_connect(self, dstaddr, dstport, *, net, timeout_factor, supports_v2_p2p=False):
```
nice! done. removed default value for `reconnect` in `peer_accept_connection` too since that is also always passed from `add_outbound_p2p_connection`

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -321,6 +321,8 @@ def _on_data(self):
                         return
                     self.recvbuf = self.recvbuf[msglen:]
 
+                    if msg is None:  # reject decoy messages
```
done.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,17 +4,280 @@
+                    if length == -1:
+                        return processed_length, False
+                    elif length == 0:
+                        return 0, True
```
nice catch! done. i've changed `authenticate_handshake` function a bit as discussed in https://github.com/bitcoin/bitcoin/pull/24748#discussion_r1447657783.

> I think it would be even cleaner to change v2_receive_packet to also return contents = None in case no complete packet was received. In that case, the whole elif length == 0: branch can go away.

i think we need the `elif length == 0:` now so that we can exit out of the `while not self.tried_v2_handshake:` loop in case we don't have sufficient bytes and need more bytes to be received on `recvbuf`?

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,17 +4,280 @@
+    def v2_receive_packet(self, response, aad=b''):
+        """Decrypt a BIP324 packet
+
+        Returns:
```
much simpler! done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -199,8 +211,13 @@ def connection_made(self, transport):
         logger.debug("Connected & Listening: %s:%d" % (self.dstaddr, self.dstport))
         self._transport = transport
+        # in an inbound connection to the TestNode, P2PConnection is the initiator. [TestNode <---- P2PConnection]
+        # ellswift is generated and sent immediately to begin the initial v2 handshake.
```
done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -231,12 +231,66 @@ def connection_lost(self, exc):
+            if not self.v2_state.initiating and not self.v2_state.sent_garbage:
+                # if the responder hasn't sent garbage yet, the responder is still reading ellswift bytes
+                send_handshake_bytes = self.v2_state.respond_v2_handshake(BytesIO(self.recvbuf))
+                length = len(self.v2_state.received_prefix)
```
makes sense! I've changed `respond_v2_handshake` and `complete_v2_handshake` to return the bytes consumed info too.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -231,12 +231,66 @@ def connection_lost(self, exc):
+        assert self.v2_state.peer
+        # at least 16 bytes garbage terminator and 20 bytes empty transport version packet
+        # is required to authenticate v2 handshake
+        if len(self.recvbuf) < 16 + 20:
```
true! done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -159,6 +162,9 @@ def __init__(self):
         # Should only call methods on this from the NetworkThread, c.f. call_soon_threadsafe
         self._transport = None
+        self.v2_state = None  # EncryptedP2PState object needed for v2 p2p connections
+        self.supports_v2_p2p = False  # set if the connection supports v2 p2p
```
oh i like this! done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -199,8 +211,13 @@ def connection_made(self, transport):
+            self.send_raw_message(send_handshake_bytes)
         if self.on_connection_send_msg:
-            self.send_message(self.on_connection_send_msg)
+            self.queue_messages.append(self.on_connection_send_msg) if self.supports_v2_p2p else self.send_message(self.on_connection_send_msg)
```
hmm you're right. i used the idea of a queue initially because i liked the abstraction implied by queuing unsent messages to be sent after v2 handshake/reconnection.

since `on_connection_send_msg` is always the version message, i've changed the code to work with not clearing `on_connection_send_msg` before v2 handshake/reconnection is complete.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -159,6 +162,9 @@ def __init__(self):
         self._transport = None
+        self.v2_state = None  # EncryptedP2PState object needed for v2 p2p connections
+        self.supports_v2_p2p = False  # set if the connection supports v2 p2p
+        self.queue_messages = []  # queue messages to send after initial v2 handshake
```
removed this variable as discussed in https://github.com/bitcoin/bitcoin/pull/24748#discussion_r1460467897.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -231,12 +231,66 @@ def connection_lost(self, exc):
+    # v2 handshake method
+    def v2_handshake(self):
+        """v2 handshake performed before P2P messages are exchanged (see BIP324). P2PConnection is an initiator
+        (in inbound connections to TestNode) and responder(in outbound connections from TestNode).
```
done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -231,12 +231,66 @@ def connection_lost(self, exc):
 
+    # v2 handshake method
+    def v2_handshake(self):
+        """v2 handshake performed before P2P messages are exchanged (see BIP324). P2PConnection is an initiator
```
done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -231,12 +231,66 @@ def connection_lost(self, exc):
+            * responder using `respond_v2_handshake()`, `complete_handshake()` and `authenticate_handshake()`
+
+        `initiate_v2_handshake()` is immediately done by the initiator when the connection is established in
+        `connection_made()`. the rest of the initial v2 handshake functions are handled here.
```
done.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -231,12 +231,66 @@ def connection_lost(self, exc):
+        `connection_made()`. the rest of the initial v2 handshake functions are handled here.
+        """
+        if not self.v2_state.peer:
+            length = 0
```
makes sense! i've changed `respond_handshake` and `complete_handshake` to return length consumed too.

stratospher commented on test/functional/p2p_v2_encrypted.py:
```diff
@@ -0,0 +1,123 @@
+
+            if i:
+                # check if node1 connected to node0 (but not to node0's p2p connection directly)
+                # gets blocks produced by node0's p2p connection
```
yes! i've simplified the comments. hopefully it's better now?

sr-gi replied:
Yes! This reads much clearer now :)

naumenkogs reviewed (commented):
Some low-level feedback for now, mostly nits. I will do high-level review against the BIP shortly.

sr-gi reviewed (commented):
Approach ACK

I think this looks good overall, I left some potential improvements/questions but feel free to disregard them given this is just for testing

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,17 +4,285 @@
+
+    def respond_v2_handshake(self, response):
+        """Responder begins the v2 handshake by sending its ellswift bytes and garbage. However, the responder
+        sends this after having received at least one byte that mismatches 16-byte v1_prefix.
```
i took the [text from the BIP](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki#cite_note-handshake_progress_19) here. On the python side, [`data_received()` asyncio function](https://docs.python.org/3/library/asyncio-protocol.html#asyncio.Protocol.data_received) only gets called when a bytes object is received.

(thinking about the c++ side)

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,17 +4,285 @@
+            if not byte:
+                return len(self.received_prefix), b""
+            self.received_prefix += byte
+            if self.received_prefix[-1] != v1_prefix[len(self.received_prefix) - 1]:
```
`respond_v2_handshake()` is supposed keep reading bytes from `bitcoind` and shoot out ellswift bytes when first mismatch from 16 bytes `V1_PREFIX` happens. Let's say the mismatch happens after reading 3 bytes - we still need to store the 3 bytes read somewhere (currently stored in `self.received_prefix`) since it's actually the first 3 bytes of ellswift which `bitcoind` was sending us.

we need this 3 bytes later in `complete_handshake()` to read the remaining 64-3 bytes and to compute ellswift which bitcoind sent us.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -221,14 +223,15 @@ def connection_made(self, transport):
             send_handshake_bytes = self.v2_state.initiate_v2_handshake()
             self.send_raw_message(send_handshake_bytes)
-        if self.on_connection_send_msg and not self.supports_v2_p2p:
+        if self.on_connection_send_msg and not self.supports_v2_p2p and not self.reconnect:
```
(originally because `on_connection_send_msg` would be end up being unused in v2 connection anyways and wanted to keep the `on_connection_send_msg` send during the v1 reconnection phase.)

interesting question! you're talking about a less common scenario where initiator node is preparing for v2 connection thinking responder node is v2 but v1 version message from responder node gets sent before v2 ellswift bytes from initiator node. if this happens the initiator node wouldn't attempt a reconnection using v1 since it's [receive buffer is not empty](https://github.com/bitcoin/bitcoin/blob/e69796c79c0aa202087a13ba62d9fbcc1c8754d4/src/net.cpp#L1525).

naumenkogs replied:
I think it deserves a comment :)

mzumsande replied:
The python code only supports the "reconnection" use case in one direction: TestNode connects to the P2PConnection with v2, and reconnects with v1 without anything breaking (e.g. timeouts) on the python side. The reverse direction (P2PConnection actively does reconnections itself for outgoing connections) is not supported and I don't think it needs to be. So P2PConnection is receiving an inbound connection, and here it must wait to send `on_connection_send_msg` with sending until the version is received so it doesn't send it out in the v2 attempt. Maybe a comment could just point that in the reconnection case, `on_connection_send_msg` is sent in `on_version`, not here.

I don't understand why in the existing v1 master code, the test framework, when receiving an inbound connection, sends out the version message in `connection_made` instead of doing it in `on_version` (similar to the way it is done in bitcoind, see [here](https://github.com/bitcoin/bitcoin/blob/207220ce8b767d8efdb5abf042ecf23d846ded73/src/net_processing.cpp#L3429-L3432)). After all, the usual protocol is that the initiator sends its version, and the responder processes it and responds. But that is unrelated to this PR.

stratospher replied:
> The reverse direction (P2PConnection actively does reconnections itself for outgoing connections) is not supported and I don't think it needs to be.

+1 since we only really care if TestNode/bitcoind (and not P2PConnection) does reconnection logic.

> I don't understand why in the existing v1 master code, the test framework, when receiving an inbound connection, sends out the version message in connection_made instead of doing it in on_version (similar to the way it is done in bitcoind, see [here](https://github.com/bitcoin/bitcoin/blob/207220ce8b767d8efdb5abf042ecf23d846ded73/src/net_processing.cpp#L3429-L3432)). After all, the usual protocol is that the initiator sends its version, and the responder processes it and responds. But that is unrelated to this PR.

i guess support for P2PConnection receiving an inbound connection/TestNode initiating outbound connections was only added later in the test framework when `addconnection` RPC was introduced. Also the concept of who's initiating wasn't relevant until now and we don't have access to that information in the inside functions.

maybe we can move [`initiating`](https://github.com/bitcoin/bitcoin/blob/fa4d9d5b39d5652de1d83564f19b1a0749863b16/test/functional/test_framework/v2_p2p.py#L67) from `class EncryptedP2PState` to `class P2PConnection` instead.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -205,27 +218,82 @@ def connection_made(self, transport):
         """asyncio callback when data is read from the socket."""
         if len(t) > 0:
             self.recvbuf += t
+            if self.supports_v2_p2p and not self.v2_state.tried_v2_handshake:
```
hmm other than the ellswift bytes, `v2_handshake()` also needs to read garbage bytes (could be received as separate chunks with size < 16 bytes) and other things like garbage terminator, decoy messages and version packet. i'm leaning towards keeping the interface generic and processing as we receive bytes.

sr-gi replied:
Oh, I thought the `recvbuf` was a growing-only buffer that was not cleared until the handshake was over.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -231,12 +231,61 @@ def connection_lost(self, exc):
+                length, send_handshake_bytes = self.v2_state.respond_v2_handshake(BytesIO(self.recvbuf))
+                self.recvbuf = self.recvbuf[length:]
+                if send_handshake_bytes == -1:
+                    self.supports_v2_p2p = False
```
good catch! done.

stratospher commented on src/rpc/net.cpp:
```diff
@@ -401,11 +402,16 @@ static RPCHelpMan addconnection()
 
-    const bool success = connman.AddConnection(address, conn_type);
+    if (use_v2transport && !(connman.GetLocalServices() & NODE_P2P_V2)) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Error: v2transport requested but not enabled (see -v2transport)");
```
done. liked the simpler language.

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -296,23 +297,49 @@ def _on_data(self):
+                    assert msg  # application layer messages(which aren't decoy messages) are non-empty
+                    shortid = msg[0]
+                    if shortid == 0:
+                        # a string command
```
added more comments. meant it as the [ASCII message type](https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki#v2-bitcoin-p2p-message-structure) like in v1.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -0,0 +1,283 @@
+        peer = {}
+        salt = b'bitcoin_v2_shared_secret' + MAGIC_BYTES[self.net]
+        for name, length in (('initiator_L', 32), ('initiator_P', 32), ('responder_L', 32), ('responder_P', 32),
+                             ('garbage_terminators', 32), ('session_id', 32)):
```
true! removed.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -0,0 +1,283 @@
+            peer[name] = hkdf_sha256(salt=salt, ikm=ecdh_secret, info=name.encode('utf-8'), length=length)
+        peer['initiator_garbage_terminator'] = peer['garbage_terminators'][:16]
+        peer['responder_garbage_terminator'] = peer['garbage_terminators'][16:]
+        del peer['garbage_terminators']
```
nice! done.

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -0,0 +1,283 @@
+    def __init__(self, *, initiating, net):
+        self.initiating = initiating  # True if initiator
+        self.net = net
+        self.peer = {}  # object with various BIP324 derived keys and ciphers
```
right, not much of a difference - leaving it as a follow up if desired.

naumenkogs commented on test/functional/p2p_v2_earlykeyresponse.py:
```diff
@@ -0,0 +1,85 @@
+from test_framework.v2_p2p import EncryptedP2PState
+
+
+class TestEncryptedP2PState(EncryptedP2PState):
```
The goal of this class is mostly to test the tests, right? Commenting that more clearly could help.

stratospher replied:
done.

naumenkogs commented:
ACK fa4d9d5b39d5652de1d83564f19b1a0749863b16

There are many more tests possible, right? E.g. sending more than 4095 bytes of garbage.
I assume the goal here wasn't to cover everything?

stratospher commented:
> There are many more tests possible, right? E.g. sending more than 4095 bytes of garbage.
I assume the goal here wasn't to cover everything?

yes! goal here was to just introduce support for v2 in the test framework. we need more tests after this.

mzumsande reviewed (commented):
Code Review ACK bc9283c4415a932ec1eeb70ca2aa4399c80437b3

theStack reviewed (approved):
Code-review ACK bc9283c4415a932ec1eeb70ca2aa4399c80437b3

Went through another review round and only found some minor nits, mostly naming suggestions. If you retouch (or follow-up material):
* should use multi-line imports (see style guidelines in `test/functional/README.md`)
* commit 382894c3acd2dbf3e4198814f547c75b6fb17706 has a leading space in the commit subject

sr-gi commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,17 +4,281 @@
+    def __init__(self, *, initiating, net):
+        self.initiating = initiating  # True if initiator
+        self.net = net
+        self.peer = {}  # object with various BIP324 derived keys and ciphers
```
I think this comes straight from BIP324 pseudocode: https://github.com/bitcoin/bips/blob/master/bip-0324.mediawiki#keys-and-session-id-derivation

naumenkogs commented:
ACK bc9283c4415a932ec1eeb70ca2aa4399c80437b3

glozow reviewed (commented):
ACK bc9283c4415a932ec1eeb70ca2aa4399c80437b3

Did some light code review of the `P2PConnection` functions, mutation testing of `EncryptedP2PState`, and manually changing other functional tests to use v2 connections.

stratospher commented:
hmm weird, looking into it.

i also have a [WIP branch](https://github.com/stratospher/bitcoin/commits/more-v2-tests/) which uses this test file and adds more v2 tests for sending excess garbage bytes, wrong garbage terminator, incorrect ellswift and non-empty version packet.

stratospher commented on test/functional/p2p_v2_earlykeyresponse.py:
```diff
@@ -0,0 +1,85 @@
+        ellswift_and_garbage_data = peer1.v2_state.initiate_v2_handshake()
+        peer1.send_raw_message(ellswift_and_garbage_data)
+        peer1.wait_for_disconnect(timeout=5)
+        self.log.info('successful disconnection when MITM happens in the key exchange phase')
```
true, will clean it up in [this branch](https://github.com/stratospher/bitcoin/commits/more-v2-tests/).

sr-gi commented:
> WIP branch

Is this consistently failing for you (or can you at least reproduce it)? I can see it also failing on some of the current open PRs but I can not get my head around why. This should only happen if `PeerEarlyKey` receives any data from the node before sending anything at all, or after sending the network magic but before sending the remaining initial bytes + garbage.

To my understanding, the node should not have been able to parse enough data to be replying (with either v1 or v2) :S

mzumsande commented:
see #29352 for a test (and a way to reproduce it).

stratospher commented on src/rpc/net.cpp:
```diff
@@ -401,11 +402,16 @@ static RPCHelpMan addconnection()
     } else {
         throw JSONRPCError(RPC_INVALID_PARAMETER, self.ToString());
     }
+    bool use_v2transport = !request.params[2].isNull() && request.params[2].get_bool();
```
done in #29356

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -590,6 +590,13 @@ def wait_for_disconnect(self, timeout=60):
 
+    def wait_for_reconnect(self, timeout=60):
+        def test_function():
+            if not (self.is_connected and self.last_message.get('version') and self.v2_state is None):
+                return False
+            return True
```
done in #29356

stratospher commented on test/functional/test_framework/p2p.py:
```diff
@@ -232,13 +232,62 @@ def connection_lost(self, exc):
         self.on_close()
 
+    # v2 handshake method
+    def v2_handshake(self):
```
done in #29356

stratospher commented on test/functional/test_framework/v2_p2p.py:
```diff
@@ -4,17 +4,281 @@
+        self.privkey_ours, self.ellswift_ours = ellswift_create()
+        garbage_len = random.randrange(MAX_GARBAGE_LEN + 1)
+        self.sent_garbage = random.randbytes(garbage_len)
+        logger.debug(f"sending {garbage_len} bytes of garbage data")
```
didn't include this in #29356 because we'd need to call it in 2 places (before `initiate_v2_handshake`, `respond_v2_handshake`) and also calculate `garbage_len` there again.

stratospher replied:
done in #29431 actually
